### DISCUSSION TITLE
Unable to derive instances of the following type class

### DISCUSSION THREAD

**Discussion by julienrf** [typelevel/shapeless-3#78] [Q&A]
I couldn’t find a way to derive instances of the following typeclass. I would like to know if this is intended by design, or if you think there is something missing in the library to support my use case.

Consider the typeclass `Record`, defined as follows:

~~~ scala
trait Records {
  type Field[A]
  type Record[A]
  extension [A] (recordA: Record[A])
    def zip[B](recordB: Record[B]): Record[(A, B)]
    def xmap[B](f: A => B)(g: B => A): Record[B]

  def field[A](name: String)(using Field[A]): Record[A]

  given int: Field[Int]
  given string: Field[String]
}
~~~

We can use it like so to manually create an instance of a data type `User`:

~~~ scala
case class User(name: String, age: Int)

trait ManualInstance extends Records {

  val user: Record[User] =
    field[String]("name")
      .zip(field[Int]("age"))
      .xmap((name, age) => User(name, age))(user => (user.name, user.age))

}
~~~

I would like to automate the derivation process: for every field `t: T` of a case class, I want to call `field[T]("t")`, and then I want to combine the records together with the `zip` method, and finally I want to transform them to a case class with `xmap`.

I wrote the following, but I am stuck:

~~~ scala
trait GenericInstances extends Records {

  def derivedRecord[A](using instances: K0.ProductInstances[Field, A]): Record[A] =
    ???

  // and then...

  derivedRecord[User]

}
~~~

It seems that none of the methods on `K0.ProductInstances` would work for me.

**Comment by joroKr21** [comment]
More missing methods: https://github.com/medeia/medeia/pull/259

We're definitely missing `construct` with applicative semantics.
But I can't tell if that would be enough for your case.
